File: m/rls/rlb/oswin.c
/* file -  oswin - Windows support */
#include "m:\rid\rider.h"
#include "m:\rid\osdef.h"
#include <stdlib.h>
#include <windows.h>
/* code -  locals */
#define esCsuc  0
#define esCwar  0
#define esCerr  1
#define esCfat  1
#if Wnt
int osVsys = osKw32;
int osVimp = osKwnt;
int osVhst = osKwnt;
#else 
int osVsys = osKw16;
int osVimp = osKwin;
int osVhst = osKdos;
#endif 
int osVcpu = osKx86;
int osVend = osKbig;
int osVsev = 0;
/* code -  os_ini - init for o/s */
os_ini()
{ return 1;
} 
/* code -  os_war - register warning */
void os_war()
{ if ( osVsev < esCwar) {osVsev = esCwar ;}
} 
/* code -  os_err - register error */
void os_err()
{ if ( osVsev < esCerr) {osVsev = esCerr ;}
} 
/* code -  os_fat - register fatal error */
void os_fat()
{ osVsev = esCfat;
} 
/* code -  os_exi - exit image */
os_exi()
{ 
#if Wnt
  exit ( osVsev);
#else 
  exit ( osVsev);
#endif 
  return 0;
} 
/* code -  os_idl - execute system idle loop */
os_idl()
{ 
  return 1;
} 
/* code -  os_wai - wait n milliseconds */
os_wai(
ULONG cnt )
{ int min ;
  if( !cnt)return 1;
  while ( cnt--) {
    min = 1000000;
    while ( --min) {;}
  } 
  return 1;
} 
/* code -  rt_ctc - zortech code */
#ifdef __SC__
volatile int rtFctc = 0;
int rtFres = 0;
/* code -  rt_ctc - catch ctrl/c */
int rt_ctc()
{ ++rtFctc;
  return ( 0);
} 
/* code -  rt_ini - init system stuff */
rt_ini()
{ 
} 
/* code -  rt_exi - exit cleanup */
rt_exi()
{ 
} 
#endif

File: m/rls/oswin.r
file	oswin - Windows support
include rid:rider
include	rid:osdef
include	rid:wimod
include	rid:dbdef
include	<stdlib.h>			; exit()

data	locals

	esCsuc := 0			; success
	esCwar := 0			; warning (treated as success)
	esCerr := 1			; error
	esCfat := 1			; fatal

If Wnt
	osVsys : int = osKw32		; system
	osVimp : int = osKwnt		; implementation
	osVhst : int = osKwnt		; host
Else
	osVsys : int = osKw16		; system
	osVimp : int = osKwin		; implementation
	osVhst : int = osKdos		; host
End
	osVcpu : int = osKx86		; cpu
	osVend : int = osKbig		; endian
	osVsev : int = esCsuc		; error severity

code	os_ini - init for o/s

  func	os_ini
  is	fine
  end

code	os_war - register warning

  proc	os_war
  is	osVsev = esCwar if osVsev lt esCwar
  end

code	os_err - register error

  proc	os_err
  is	osVsev = esCerr if osVsev lt esCerr
  end

code	os_fat - register fatal error

  proc	os_fat
  is	osVsev = esCfat
  end

code	os_exi - exit image

  func	os_exi 
  is
If Wnt
	exeunt osVsev
;sic]	ExitProcess (osVsev)		; aborts I/O without rundown
Else
	exeunt osVsev
End
	fail
  end

code	os_idl - execute system idle loop

;	Duration is undefined

  func	os_idl
  is	fine Sleep (0)		; give up rest of timeslice
  end

code	os_wai - wait n milliseconds

;	Zero returns immediately
;	~0 may be infinite

  func	os_wai
	cnt : LONG
  is	min : int
	fine if !cnt			; waits no time at all
	fine Sleep (cnt)		; portable wait
;	while cnt--
;	   min = 1000000		;
;	   spin while --min
;	end
;	fine
  end

code	os_w95 - is Windows 95

  func	os_w95
  is	reply  (GetVersion () & 0xc0000000) eq 0xc0000000
  end

code	os_wnt - is Windows Nt

  func	os_wnt
  is	reply !os_w95 ()
  end

code	os_dbg - is debug (not retail)

  func	os_dbg
  is	reply GetSystemMetrics (SM_DEBUG) ne
  end
code	rt_ctc - zortech code
#ifdef __SC__
;include <int.h>			; signals

;	type rtTint : INT_DATA struct	; interrupt frame
	rtFctc : int volatile = 0	; ctrl/c seen
	rtFres : int = 0		; need to restore vector

code	rt_ctc - catch ctrl/c

  func	rt_ctc 				; signal routine
;	frm : * rtTint			; interrupt frame
	() : int			; 0 => call previous
  is	++rtFctc			; remember it
	reply 0				; call previous handler
  end

code	rt_ini - init system stuff

  func	rt_ini
  is
;	int_intercept (0x1b, &rt_ctc, 200)
;	++rtFres if that eq		; succeeded - restore later
  end

code	rt_exi - exit cleanup

  func	rt_exi
  is
;	if rtFres			; need to restore
;	.. int_restore (0x1b)		; restore break vector
  end
#endif
code	os_del - delete file

  func	os_del
	spc : * char
	sts : * int
  is	fine if DeleteFile (spc)	; delete worked
	fail if !sts			; no status required
;	err = GetLastError ()		; find out why
;	pass fine
;	db_lst ("Delete")
	fail
  end
